### PR TITLE
add support for sharding mongodb collections

### PR DESCRIPTION
**Pull request by rodesai**
Adds support for sharding mongodb collections. Sharding is enabled using a config property

### CODE REVIEW

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWriter.java [hunk 1]
```diff
@@ -69,8 +70,9 @@ public CompletionStage<RemoteWriteResult<P>> flush() {
       return CompletableFuture.completedFuture(RemoteWriteResult.success(tablePartition));
     }
 
+    final BulkWriteOptions options = new BulkWriteOptions().ordered(false);
```
Comment (rodesai) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWriter.java:73]: this change is unrelated to sharding, but we need it for flush performance

File: kafka-client/src/main/java/dev/responsive/kafka/api/config/ResponsiveConfig.java [hunk 2]
```diff
@@ -165,6 +165,16 @@ public class ResponsiveConfig extends AbstractConfig {
       + "with for best results. However it is important to note that this cannot be changed for "
       + "an active application. Messing with this can corrupt existing state!";
 
+  public static final String MONGO_COLLECTION_SHARDING_ENABLED = "responsive.mongo.collection.sharding.enabled";
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/config/ResponsiveConfig.java:168]: nit: for consistency 
```suggestion
  public static final String MONGO_COLLECTION_SHARDING_ENABLED_CONFIG = "responsive.mongo.collection.sharding.enabled";
```

File: kafka-client/src/main/java/dev/responsive/kafka/api/config/ResponsiveConfig.java [hunk 3]
```diff
@@ -165,6 +165,16 @@ public class ResponsiveConfig extends AbstractConfig {
...
+  private static final boolean MONGO_COLLECTION_SHARDING_ENABLED_DEFAULT = false;
+  private static final String MONGO_COLLECTION_SHARDING_ENABLED_DOC = "Toggles use of sharded collections. Set"
+      + "this to true when running against a sharded mongo cluster, to shard a collection across multiple mongo"
+      + "replica sets.";
+
+  public static final String MONGO_COLLECTION_SHARDING_CHUNKS = "responsive.mongo.collection.sharding.chunks";
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/config/ResponsiveConfig.java:174]:
```suggestion
  public static final String MONGO_COLLECTION_SHARDING_CHUNKS_CONFIG = "responsive.mongo.collection.sharding.chunks";
```

File: kafka-client/src/main/java/dev/responsive/kafka/api/config/ResponsiveConfig.java [hunk 4]
```diff
@@ -165,6 +165,16 @@ public class ResponsiveConfig extends AbstractConfig {
       + "with for best results. However it is important to note that this cannot be changed for "
       + "an active application. Messing with this can corrupt existing state!";
 
+  public static final String MONGO_COLLECTION_SHARDING_ENABLED = "responsive.mongo.collection.sharding.enabled";
+  private static final boolean MONGO_COLLECTION_SHARDING_ENABLED_DEFAULT = false;
+  private static final String MONGO_COLLECTION_SHARDING_ENABLED_DOC = "Toggles use of sharded collections. Set"
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/config/ResponsiveConfig.java:170]: super minor nit, but all the line breaks are missing spaces

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java [hunk 5]
```diff
@@ -171,6 +194,7 @@ public MongoWindowedTable(
     );
 
     this.database = client.getDatabase(name).withCodecRegistry(pojoCodecRegistry);
+    this.adminDatabase = client.getDatabase("admin");
     this.metadata = database.getCollection(
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java:198]: We intentionally only shard the per-partition data collections, right? ie we don't need to shard the shared metadata collection?

**Review by ableegoldman — COMMENTED** [review]
Few small things but overall LGTM

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java (hunk 5 above)
Comment (rodesai) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java:198]: Now that you mention it I suppose we could. Seems simpler to just leave it unsharded in the sense that there are less possible configurations for it.

**Review by ableegoldman — APPROVED** [review]
LGTM

**Pull request merged**